antlr4-cppruntime: Upgrade to 4.11.0/4.11.1

0xFireWolf opened the pull request:
Specify library name and version:  **antlr4-cppruntime/4.11.0, 4.11.1**

This PR contains the following changes:
- Upgrade the runtime library to 4.11.0 and 4.11.1.
- Drop the dependency on `libuuid` as of 4.11.0.
    - Related PR: https://github.com/antlr/antlr4/pull/3787/
- Support ARM architectures as of 4.11.0.
    - The previous receipt mentions that ARM is not supported due to the missing library `libuuid`.
- ~~The minimum version of GCC is now 8 (previously 7) to target C++17.~~ (Will be reverted; See Below.)

Thank you.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

0xFireWolf commented:
The compiler errors reported by GCC 7 has been fixed by the PR https://github.com/antlr/antlr4/pull/3885 which is pending review and approval. The minimum compiler requirement will be restored from GCC 8 to GCC 7. Depending upon when the next version of antlr4 will be released, I may choose to include a patch file that fixes the compiler errors to ship `antlr4-cppruntime/4.11.1` as soon as possible.

----

- [ ] Antlr4 Pull request has been merged
- [ ] The compiler requirement has been restored
- [ ] The patch file specified to 4.11.0/4.11.1 is now included

danimtb commented:
there are some annotations made by the linter to support conan v2. @0xFireWolf would you mind having a look at them? thank you!

0xFireWolf commented:
> there are some annotations made by the linter to support conan v2. @0xFireWolf would you mind having a look at them? thank you!

It would be great if we can merge this PR as soon as possible.
Unfortunately, I don't have time for making this recipe compatible with Conan v2 at this moment, as I am still not quite familiar with writing the recipe.

danimtb commented:
@0xFireWolf no problem. Unfortunately we cannot disable the checks, but don't worry, I will be taking over this PR if you don't mind 😄

uilianries commented:
@0xFireWolf please, consider https://github.com/0xFireWolf/conan-center-index/pull/1

0xFireWolf commented:
> @0xFireWolf please, consider [0xFireWolf#1](https://github.com/0xFireWolf/conan-center-index/pull/1)

Thanks for the PR. I really appreciate it.

0xFireWolf commented:
> @0xFireWolf no problem. Unfortunately we cannot disable the checks, but don't worry, I will be taking over this PR if you don't mind 😄

Please feel free to do so. Thank you.

uilianries commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +29,129 @@ class Antlr4CppRuntimeConan(ConanFile):
-        antlr_version = tools.Version(self.version)
-
-        if compiler == "Visual Studio" and compiler_version < "16":
+        if (self.info.settings.compiler == "Visual Studio" and self.info.settings.compiler.version < "16") or \
+           (self.info.settings.compiler == "msvc" and self.info.settings.compiler.version < "1920"):
```
```suggestion
        if (self.info.settings.compiler == "Visual Studio" and Version(self.info.settings.compiler.version) < "16") or \
           (self.info.settings.compiler == "msvc" and Version(self.info.settings.compiler.version) < "1920"):
```

Need to parse version

michalwidera commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +29,129 @@ class Antlr4CppRuntimeConan(ConanFile):
             # Version 16 compiles this code correctly.
 
-        if antlr_version >= "4.10":
+        if Version(self.version) >= "4.10":
```
Are you sure that Version(self.version) is same tools.Version(self.version) ?

uilianries commented:
I tried locally and I got:

```
-- Checking for module 'uuid'
--   Package 'uuid', required by 'virtual:world', not found
CMake Error at /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:545 (message):
 A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/FindPkgConfig.cmake:733 (_pkg_check_modules_internal)
  CMakeLists.txt:44 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
See also "/home/conan/.conan/data/antlr4-cppruntime/4.9.3/_/_/build/4d27e06017598e295f7b1604bc676563db9ef0f6/build/Release/CMakeFiles/CMakeOutput.log".

```

The point is, libuuid is not prepared to generate PkgConfigDeps: https://github.com/conan-io/conan-center-index/blob/master/recipes/libuuid/all/conanfile.py#L78

Even the legacy generator "pkg_config" is incompatible with new generators, so mixing them will not work. Will open a PR updating Libuuis asap.

uilianries commented:
@0xFireWolf I just created the PR https://github.com/conan-io/conan-center-index/pull/13155

michalwidera commented:
Please rebase (#13130 has been merged) and try to rebuild.

uilianries commented:
Do not to rebase anything, just start a new CI build. Any dependency is consumed as a Conan package, git status does not matter for conan commands (which is very good!)

uilianries commented:
Just starter a new CI job now 🤞

0xFireWolf commented:
It seems that `libuuid` with `PkgConfigDeps` is still problematic.

The PR (https://github.com/antlr/antlr4/pull/3787) I mentioned above indicates that `libuuid` is not used at least since 4.9, so I have created patches that drop it from 4.9.3, 4.10 and 4.10.1 and enable ARM builds on those versions. As such, the runtime library solely relies on `utfcpp` which is a header-only library. 

I also remove the `PkgConfigDeps` generator from the Conan file, because I don't think it is needed any more. Let's see whether the CI build will succeed or not.

michalwidera commented:
If I understand the cause of build failure there is a problem with test code / validation.
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/12902/11-configs/windows-visual_studio/antlr4-cppruntime/4.9.3//summary.json
Failed: BUILT + TEST_FAILED on msvc 16 with following core issue:
```
error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in test_package.obj
```
Therefore test artifact didn't appear and this stopped validation of this update.

michalwidera commented on recipes/antlr4-cppruntime/all/test_package/CMakeLists.txt:
```diff
@@ -12,4 +9,4 @@ if(TARGET antlr4_shared)
     target_link_libraries(${PROJECT_NAME} PRIVATE antlr4_static)
 endif()
-set_target_properties(${PROJECT_NAME} PROPERTIES CXX_STANDARD 17 CXX_STANDARD_REQUIRED ON)
+target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)
```
This could be a cause of error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MTd_StaticDebug' doesn't match value 'MDd_DynamicDebug' in test_package.obj
ref:
https://stackoverflow.com/questions/14714877/mismatch-detected-for-runtimelibrary

0xFireWolf replied:
I think the linker error might be related to this PR: [Windows specific fixes](https://github.com/antlr/antlr4/pull/3678).
The CI result shows that all tests passed on Windows with antlr 4.11.x.

0xFireWolf replied:
Now I am curious about why this error didn't show up when the recipe was upgraded to support 4.10.x last time and what the best way to solve this error this time is (i.e. modifying the Conan file versus backporting the fix). This linker error showed up after I merged @uilianries's PR that made the recipe compatible with Conan 2.0, so maybe he can shed some lights on this issue.

michalwidera replied:
You talk about DropLibUuid-4.9.3 patchset ? That was @uilianries change ?

uilianries replied:
Nothing to do with Conan 2.0 neither libuuid. Please, look into upstream CMakelists and check for static runtime forced flags. It will require a patch and if already available, please, add to conandata.yml all information.

0xFireWolf commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,118 @@ class Antlr4CppRuntimeConan(ConanFile):
-        cmake = self._configure_cmake()
+        apply_conandata_patches(self)
+        cmake = CMake(self)
+        cmake.configure(build_script_folder="runtime/Cpp")
```
Finally, I solved the linker errors when building 4.9.3 and 4.10.x with the Conan 2.0 recipe.
This "bug" in antlr's `CMakeLists.txt` was revealed after I merged [uilianries](https://github.com/uilianries)'s PR.

The CMake variable `WITH_STATIC_CRT` that controls whether `antlr` should be linked against the static MSVC runtime does not change at all. The problem is that we need to use a `CMakeLists.txt` wrapper to configure CMake, otherwise it would be too late to set that variable, and as a result `antlr` is always linked against the static MSVC runtime.

0xFireWolf replied:
I can build the package on Windows without any errors now. 
Let's see whether the CI build will succeed or not this time.

0xFireWolf commented:
All green. @danimtb @uilianries 
It would be great if we can merge this as soon as possible. 
Thank you.

---

**Summary of this PR:**
- Migrated the recipe to Conan 2.0 (A huge thanks to @uilianries).
- Removed the dependency on `libuuid` from 4.9.3 and 4.10.x.
- Fixed the linker errors related to MSVC runtime revealed by the 2.0 recipe.
- Added support for ARM builds for 4.9.3, 4.10.x and 4.11.x.
- Fixed the compiler errors reported by GCC 7 due to the new `FlatHashSet` and `FlatHashMap` types introduced in 4.11.x.

michalwidera commented:
0xFireWolf - I've asked for reviewers on Slack Cpplang/conan channel.

uilianries commented on recipes/antlr4-cppruntime/all/CMakeLists.txt:
```diff
@@ -1,7 +1,4 @@
-conan_basic_setup(KEEP_RPATHS)
-
-add_subdirectory(source_subfolder/runtime/Cpp)
+add_subdirectory(runtime/Cpp)
```
This file is no longer needed, please remove it. Now, CMakeToolchain does all the magic.

uilianries commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,126 @@ class Antlr4CppRuntimeConan(ConanFile):
-        self.copy("CMakeLists.txt")
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            self.copy(patch["patch_file"])
+        # We need this wrapper to build the library on Windows properly.
```
Why don't you set CMakeToolchain.variable["WITH_STATIC_CRT"] instead?

prince-chrismc replied:
You can use ``cache_variables`` if the project does not respect the CMake variables

uilianries commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,126 @@ class Antlr4CppRuntimeConan(ConanFile):
-        antlr_version = tools.Version(self.version)
-
-        if compiler == "Visual Studio" and compiler_version < "16":
+        if (compiler == "Visual Studio" and compiler_version < "16") or \
+           (compiler == "msvc" and compiler_version < "1920"):
```
It could be very simplified by using `check_min_vs_("1920")` 
https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html#check-min-vs

uilianries commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,126 @@ class Antlr4CppRuntimeConan(ConanFile):
+        tc.generate()
+        tc = CMakeDeps(self)
+        tc.generate()
+        tc = VirtualBuildEnv(self)
+        tc.generate()
```
```suggestion
```
As there is no build_requirement anymore, it's no longer needed.

prince-chrismc commented on recipes/antlr4-cppruntime/all/conandata.yml:
```diff
@@ -8,7 +8,26 @@ sources:
   "4.9.3":
     - patch_file: "patches/0001-update-cmakelist.patch"
-      base_path: "source_subfolder"
+    - patch_file: "patches/0004-update-DropLibuuid-4.9.3-1.patch"
```
please add the metadata  patch fields so we can track these better 🙏 

https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

I would also suggestion opening a PR upstream with the fixes

prince-chrismc commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,126 @@ class Antlr4CppRuntimeConan(ConanFile):
+        # As of 4.11, antlr4-cppruntime no longer requires libuuid.
+        # Reference: [C++] Remove libuuid dependency (https://github.com/antlr/antlr4/pull/3787)
+        # Note that the above PR points that libuuid can be removed from 4.9.3, 4.10 and 4.10.1 as well.
+        # We have backport the fix to these old versions.
```
This does not sound like it matches our patching policy 

https://github.com/conan-io/conan-center-index/blob/master/docs/policy_patching.md

You should preserve the behavior of the release as-the author intended... back-porting features will not match the consumers expections

0xFireWolf replied:
I think this patch is fairly important because
- `libuuid` is not used since at least ANTLR 4.9 (revealed in https://github.com/antlr/antlr4/pull/3787).
- `libuuid` prevents compilation on Apple Silicon-based Macs (notes added by @michalwidera).
- `libuuid` requires the `PkgConfig` generator which is not working when @uilianries migrated the recipe to Conan 2.0. 
  - @uilianries's comment: https://github.com/conan-io/conan-center-index/pull/12902#issuecomment-1257790435
  - The failure reported by CI after `libuuid` was made compatible with Conan 2.0: https://github.com/conan-io/conan-center-index/pull/12902#issuecomment-1263653241.

As such, I decided to backport this fix from 4.11.x to 4.9.3 and 4.10.x because doing so makes `antlr` no longer dependent on non-header-only libraries and **does not change `antlr`'s runtime behavior at all**.

prince-chrismc commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +28,126 @@ class Antlr4CppRuntimeConan(ConanFile):
-        if self.settings.os in ("FreeBSD", "Linux"):
-            self.build_requires("pkgconf/1.7.4")
+            if is_msvc(self) and antlr_version == "4.10":
+                raise ConanInvalidConfiguration(f"{self.ref} is broken on msvc - Use 4.10.1 or above.")
```
Lets just not package it 🤔 no point in added a dead release

0xFireWolf replied:
Good. I will remove both 4.10.0 and 4.11.0 because they were both dead.

prince-chrismc commented on recipes/antlr4-cppruntime/all/patches/0004-update-DropLibuuid-4.10.x-1.patch:
```diff
@@ -0,0 +1,13 @@
+ 
+-if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+-  find_package(PkgConfig REQUIRED)
+-  pkg_check_modules(UUID REQUIRED uuid)
```
How can this be removed without changing the code? 😕

0xFireWolf replied:
This PR [[C++] Remove libuuid dependency](https://github.com/antlr/antlr4/pull/3787) showed that `libuuid` is not used since at least ANTLR 4.9. You can find the conclusion at here: https://github.com/antlr/antlr4/pull/3787#issuecomment-1206106174.

0xFireWolf commented:
TODO:
- [ ] Improve the recipe to deal with linker errors on Windows elegantly (4.9.3 and 4.10.x).
    - [ ] Give a try to set the cache variables for CMake.
    - [ ] Get rid of the `CMakeLists.txt` wrapper.
- [ ] Use the new function `check_min_vs()` to handle both `Visual Studio` and `msvc` easily.
- [ ] Remove the `VirtualBuildEnv` toolchain because there is no more build requirement.
- [ ] Drop the version 4.10.0 and 4.11.0 because they were immediately superseded by their first minor release.
- [ ] Add the metadata for each patch for better record keeping.
- [ ] State the minimum C++ standard for each antlr version more clearly.
    - [ ] Consider to refactor the `validate()` function further to check the compiler thoroughly. (i.e. Version < 4.10 requires C++11 while Version >= 4.10 requires C++17. There is no check against C++11 at this moment.)

0xFireWolf commented:
Since I am not the original author of this recipe, @michalwidera please feel free to chime in and add your suggestions to make this recipe better. : )

michalwidera commented:
> Since I am not the original author of this recipe, @michalwidera please feel free to chime in and add your suggestions to make this recipe better. : )

First of all - topic: 'author' - I'm not binded to this code, this is open source - anyone can change this code - there is no 'mine' code here. I don't feel that someting belongs to me here. This is common good. I've only give small contribution. Anyone who have skill and thinks that there is better way, go ahead! Please improve - there is no 'my intention' here. Only better code is leading rule. 
I'm very happy that someone else is taking care about this code - someone need's it! me either!

If we can kick off uuid - please do! If we can compile on ARM please do! That's what everyone needs!
In case of dropping 4.10.0 and 4.11.0 I can agree with conan reviewers. 
BTW, I believe that our mistakes are not disapearing - anyway ...

Reviewers please advise!

michalwidera reviewed (approved):
LGTM

michalwidera commented:
Loogin at: https://github.com/conan-io/conan-center-index/blob/master/docs/policy_patching.md 
According to [Rules/Source patches] - we need higher authority than mine to get acceptance.

jwillikers commented on recipes/antlr4-cppruntime/all/conanfile.py:
```diff
@@ -23,133 +27,113 @@ class Antlr4CppRuntimeConan(ConanFile):
+    @property
+    def _minimum_compiler_versions_cpp17(self):
+        return {
             "Visual Studio": "16",
+            "msvc": "1923",
```
Maybe use [check_min_vs](https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html?highlight=msvc#check-min-vs) to check the Visual Studio / msvc version instead of this lookup. I'm not sure that the `msvc` version will work quite right due to that fourth digit. It would need to be `192` or `193` most likely? Or for some reason maybe `192.3`? There's an example of how to do this in [here](https://github.com/conan-io/conan-center-index/blob/master/docs/package_templates/cmake_package/all/conanfile.py#L90), although with how you're using it, you'd need to use a `try...catch` block.

0xFireWolf replied:
Thanks for pointing that out. I have removed both `Visual Studio` and `msvc` from the dictionary. 

`check_min_vs` does not work with `1923` due to the fourth digit, but it turns out that I don't need to check the MSVC version in `_check_minimum_compiler_version_cpp17()`, because the invocation of `check_min_vs(self, "192")` in `validate()` already handles that.

jwillikers reviewed (changes requested):
Fantastic! One minor quibble, but otherwise looks golden.

jwillikers reviewed (approved):
LGTM!

prince-chrismc commented:
The patching being applied here is really in a grey area for me. It's a really tough call and my opinion is not the final word

- The core principle behind the patching policy is to "provide working code" and as you have so well put it, the patch

> does not change `antlr`'s runtime behavior at all

Now, looking at this conversation and how it has changes for the last month for almost 2 hours I have some suggestions and comments.

## 😻 💟 ❤️ **thank you so much for the time and patients getting this PR updated** ❤️ 💟 😻 

I absolutely agree this patch at one point could be close to the "working code" requirement. However I think a few things have changed in the last two weeks since I started reviewing this PR so I am on the other side of the fence.

Let me break down my perspective.

### Problem 1: Antlr was not using UUID

> * `libuuid` is not used since at least ANTLR 4.9 (revealed in [[C++] Remove libuuid dependency  antlr/antlr4#3787](https://github.com/antlr/antlr4/pull/3787)).
> As such, I decided to backport this fix from 4.11.x to 4.9.3 and 4.10.x because doing so makes `antlr` no longer dependent on non-header-only libraries and **does not change `antlr`'s runtime behavior at all**.

"its not being use by _this_ library" does not mean the Conan consumer is not expecting it to be present.
Conan users very often pull more than one package and given how 1.x works people expect ALL of the deps to be made available backporting this will break their expectations as the 4.9.3 did in fact require libuuid.

I absolutely understand this respects the correctness of this libary, but as a ConanCenter contributors my focus is more on the consumers who rely on the packages. How would they see this change if it broke their project?

### Problem 2: ARM Support

> * `libuuid` prevents compilation on Apple Silicon-based Macs (notes added by @michalwidera).

It does in fact work. https://packages.ubuntu.com/bionic/uuid-devhas been shipping for a while on ARM but the recipe just had not been updated to work there. The last PR published packages for it https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/13130/1/libuuid/1.0.3//summary.json so it should be availble now.

Again this was added the well after PR was started, the comments made were accurate but the speed at which we are moving is really exciting to see :rocket: so it could be reconsidered.

### Problem 3: New Generator Support

> * `libuuid` requires the `PkgConfig` generator which is not working when @uilianries migrated the recipe to Conan 2.0.
>   
>   * @uilianries's comment: [antlr4-cppruntime: Upgrade to 4.11.0/4.11.1 #12902 (comment)](https://github.com/conan-io/conan-center-index/pull/12902#issuecomment-1257790435)
>   * The failure reported by CI after `libuuid` was made compatible with Conan 2.0: [antlr4-cppruntime: Upgrade to 4.11.0/4.11.1 #12902 (comment)](https://github.com/conan-io/conan-center-index/pull/12902#issuecomment-1263653241).
> 

As for things not working with the new generators https://github.com/conan-io/conan-center-index/blob/d9ad528b30032bb01c5d522224e5b604c6fa6501/recipes/libuuid/all/conanfile.py
Is was migrated after those comments were written which is not surprising given how fast things are changing

As for the build problem that was encountered, I am pretty sure it's related to this https://github.com/conan-io/conan/issues/11962
The CMakeToolchain provided paths and the cache layout means the build-in CMake pkg_config support does not work... I tried to look but the workaround does not seem to have been used in the PR, which would certainly cause the problem that was seen.

I think it would be worth giving it a second shot with this new information. 😸 

***

I 100% think the work that was done was right, you've all done fantastic work

0xFireWolf commented:
@prince-chrismc First of all, I want to thank you for the thoughtful comments and suggestions.

**RE: Problem 1: Antlr was not using UUID**

> "its not being use by this library" does not mean the Conan consumer is not expecting it to be present. People expect ALL of the deps to be made available.

This does not make sense to me at all. If developers need `libuuid` for their projects, they should **explictly** add `libuuid/1.0.3` to their `conanfile.txt`. Importing a package implicitly should be strongly discouraged or even disallowed. To me, their expectation was actually misled or, precisely speaking, is wrong because of Conan 1.x being too permissive. The C compiler even warns the users when they use a function implicitly without including the header. 

> But as a ConanCenter contributors my focus is more on the consumers who rely on the packages. How would they see this change if it broke their project?

I duly understand that you care a lot about your consumers, but to some extent your comments make me feel like you are being overprotective of them. Yes, they will observe that their projects cannot be compiled if they import `libuuid` implicitly, but I believe that this is also their opportunity to fix this bug in their build system. We could certainly add a print statement in `requirements()`, warning them to add `libuuid` to their `conanfile.txt` if they really need it. 

To be honest, I didn't even know that `antlr` depended on `libuuid` until I upgraded this recipe, because I trust Conan to resolve all the dependencies for me automatically and correctly. I do hope that other antlr users have the same experience as me, so that we won't even notice the removal of `libuuid`.

**RE: Problem 2: ARM Support**

I sincerely thank you for the information because I don't have the time to keep track of other recipes.
However, whether the `libuuid` recipe supports `ARM` now is no longer important because it was not and will not be needed by `antlr`.

**RE: Problem 3: New Generator Support**

Your comment https://github.com/conan-io/conan/issues/11962#issuecomment-1273843499 posted 8 hours ago indicated that the workaround did not fix all the issues. As such, implementing this hacky workaround for the generator of a package which is actually not used by `antlr` at all is just a waste of effort. Besides, I believe that this discrepancy of `pkgconfig` will eventually be fixed in a newer Conan release, so are we going to remove this workaround or enable this workaround conditionally when the new Conan is released? To me, the best solution is to remove `libuuid`, solving the issue once and for all. This recipe remains concise and respects the correctness of the antlr library.

**My conclusion**

I am not inclined to add `libuuid` back to this recipe.
Antlr 4.11.1 was released on Sep 4, and I opened this PR on Sep 11, hoping to ship 4.11.1 as soon as possible.
Making the recipe compatible with Conan 2.0 was not even the goal of this PR, but thanks to uilianries's help we made it.
This update is way past due, and as one of your consumers, I am also looking forward to upgrading to antlr 4.11.1 for my projects.
Unfortunately, I do not have the time for the so-called second shot.
If we cannot reach a mutual agreement or you are uncomfortable  with my proposed changes, please feel free to close this PR or I can close it myself.

Thank you and I hope that you can understand my position.

michalwidera commented:
Suggestion of way of resolving this conflict:
Remain as it is without changing the code with patches and Limit platforms for them
* 4.9.3
* 4.10.0
* 4.10.1
* 4.11.0
* 4.11.1
and then add following binaries:
* 4.9.3-patch1
* 4.10.1-patch1
* 4.11.1-patch1
That has no dependency to libuuid and have bug fixes and is able ARM compilation.
That will create non changed artifacts and libraries with required changes?
Is this plan ok for all ?
@prince-chrismc @uilianries @0xFireWolf - what you think ?

prince-chrismc commented:
Asking the Conan team for a second opinion, they can be the tie breaker 🤞 

I am completely on the fence and my personal bias around patches is not a deciding factor 😄

prince-chrismc commented:
Okay, this is a really hard one. So thanks for working with me on it. Let me capture our reasoning as precedence

***

Our criteria for patches is "makes working software" the patches in question do not reach that (per say), they do however allow building more easily. 

Given this is aligned with the projects expectations, combined with the uncertainty of `CMakeToolchain` + `PkgConfigDeps` this is a valid backport. From the project's perspective this should have never been required and this make it more available from a Conan perspective
